contracts/src/povw/Mint.sol

```diff
@@ -0,0 +1,134 @@
+    error InvalidSteelCommitment();
+    /// @dev selector 0x82db2de2
+    error IncorrectPovwAddress(address expected, address received);
+    /// @dev selector 0xf4a2b615
+    error IncorrectInitialUpdateCommit(bytes32 expected, bytes32 received);
+
+    IRiscZeroVerifier internal immutable VERIFIER;
+    IERC20Mint internal immutable TOKEN;
+    PoVW internal immutable POVW;
+
+    // TODO(povw): Extract to a shared library along with EPOCH_LENGTH.
+    // NOTE: Example value of 100 tokens per epoch, assuming 18 decimals.
+    uint256 public constant EPOCH_REWARD = 100 * 10 ** 18;
+
+    /// @notice Image ID of the mint calculator guest.
+    /// @dev The mint calculator ensures:
+    /// * An event was logged by the PoVW contract for each log update and epoch finalization.
+    ///   * Each event is counted at most once.
+    ///   * Events form an unbroken chain from initialCommit to updatedCommit. This constitutes an
+    ///     exhaustiveness check such that the prover cannot exclude updates, and thereby deny a reward.
+    /// * Mint value is calculated correctly from the PoVW totals in each included epoch.
+    ///   * An event was logged by the PoVW contract for epoch finalization.
+    ///   * The total work from the epoch finalization event is used in the mint calculation.
+    ///   * The mint recipient is set correctly.
+    bytes32 internal immutable MINT_CALCULATOR_ID;
+
+    /// @notice Mapping from work log ID to the most recent work log commit for which a mint has occurred.
+    /// @notice Each time a mint occurs associated with a work log, this value ratchets forward.
+    mapping(address => bytes32) internal lastCommit;
+
+    constructor(IRiscZeroVerifier verifier, PoVW povw, bytes32 mintCalculatorId, IERC20Mint token) {
+        VERIFIER = verifier;
+        MINT_CALCULATOR_ID = mintCalculatorId;
+        TOKEN = token;
+        POVW = povw;
+    }
+
+    /// @notice Mint tokens as a reward for verifiable work.
+    function mint(bytes calldata journalBytes, bytes calldata seal) external {
+        // Verify the mint is authorized by the mint calculator guest.
+        VERIFIER.verify(seal, MINT_CALCULATOR_ID, sha256(journalBytes));
+        MintCalculatorJournal memory journal = abi.decode(journalBytes, (MintCalculatorJournal));
+        if (!Steel.validateCommitment(journal.steelCommit)) {
+            revert InvalidSteelCommitment();
+        }
+        if (journal.povwContractAddress != address(POVW)) {
+            revert IncorrectPovwAddress({expected: address(POVW), received: journal.povwContractAddress});
+        }
+
+        // Ensure the initial commit for each update is correct and update the final commit.
+        for (uint256 i = 0; i < journal.updates.length; i++) {
+            MintCalculatorUpdate memory update = journal.updates[i];
+
+            // On the first mint for a journal, the initialCommit should be equal to the empty root.
+            bytes32 expectedCommit = lastCommit[update.workLogId];
```

Comment:
It seems that the Mint Calculator Guest processes every single prover in the epoch, which means that for every mint there could be _a lot_ of updates + mints

How feasible is it for minting to be sybil'd? Concern would be to make mint consume so much gas that the transaction can not fit in a block, which I believe would halt rewards indefinitely (?)

Reply:
Ah, yes. So if this becomes an issue the transaction can be broken up. The segmenting can be done both between work logs (e.g. some work logs in one tx and some in another) or between updates (e.g. the first x updates in one tx and the next y updates in another tx). The smallest chunk allowed is a single `WorkLogUpdated` event for a single work log (i.e. a single prover).

In the case that there are many low value log updates, which might be so small its not worth paying the gas to mint the tokens, these can simply be excluded. If the owner of those work logs would like their mint, they can post the update themselves.

Reply:
Are you saying it can be supported in the current implementation, or its something we could do in a future upgrade? (maybe im missing it but I don't see how you can do it with the current Mint guest)